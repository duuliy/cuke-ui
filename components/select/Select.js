import React, { PureComponent, createRef } from "react";
import { createPortal } from "react-dom";
import PropTypes from "prop-types";
import cls from "classnames";
import Input from "../input";
import { DownIcon, LoadingIcon } from "../icon";
import { debounce } from "../utils";
import scrollIntoViewIfNeeded from "scroll-into-view-if-needed";
import Empty from "../empty";

const sizes = {
  default: "default",
  small: "small",
  large: "large"
};

export default class Select extends PureComponent {
  state = {
    selectedValue: this.props.defaultValue || this.props.value || "",
    visible: this.props.visible || null,
    left: 0,
    top: 0,
    width: 0
  };
  static defaultProps = {
    prefixCls: "cuke-select",
    onPanelVisibleChange: () => {},
    onChange: () => {},
    getPopupContainer: () => document.body,
    position: "bottom",
    disabled: false,
    allowClear: false,
    notFoundContent: <Empty height={120} />,
    labelInValue: false,
    loading: false
  };
  static propTypes = {
    prefixCls: PropTypes.string.isRequired,
    onPanelVisibleChange: PropTypes.func,
    getPopupContainer: PropTypes.func,
    onChange: PropTypes.func,
    disabled: PropTypes.bool,
    allowClear: PropTypes.bool,
    labelInValue: PropTypes.bool,
    size: PropTypes.oneOf(Object.values(sizes)),
    overlay: PropTypes.oneOfType([
      PropTypes.element,
      PropTypes.string,
      PropTypes.object
    ]),
    popupContainerClassName: PropTypes.string
  };
  constructor(props) {
    super(props);
    this.timeOutId = null;
    this.toggleContainer = createRef();
    this.triggerWrapper = createRef();
    this.wrapper = createRef();
  }

  static getDerivedStateFromProps(props) {
    if (Reflect.has(props, "value")) {
      return {
        selectedValue: props.value
      };
    }
    return null;
  }

  get selectedValue() {
    const { selectedValue } = this.state;
    const { labelInValue } = this.props;
    if (labelInValue) {
      // children 如果有空格 会被解析成数组 有可能是 [10,"条数据"] 所以需要 join 一下
      return Array.prototype.join.call(
        (
          this.selectOptions.find(
            ({ key }) =>
              key === ((selectedValue && selectedValue.key) || selectedValue)
          ) || {}
        ).value,
        ""
      );
    }
    return selectedValue;
  }

  get selectOptions() {
    return this.props.children.map(({ props }) => {
      return {
        key: props.value,
        value: props.children
      };
    });
  }

  onChange = (value, label) => {
    const { labelInValue } = this.props;
    this.setState({ selectedValue: value, visible: false });
    this.props.onPanelVisibleChange(false);
    if (labelInValue) {
      this.props.onChange({
        key: value,
        label
      });
    } else {
      this.props.onChange(value);
    }
  };
  onClickHandler = () => {
    const visible = !this.state.visible;
    this.setState({
      visible
    });
    this.props.onPanelVisibleChange(visible);
    if (visible) {
      this.setWrapperBounding(() => {
        scrollIntoViewIfNeeded(this.wrapper.current, {
          scrollMode: "if-needed",
          behavior: "smooth",
          block: "nearest",
          inline: "nearest"
        });
      });
    }
  };
  onClickOutsideHandler = e => {
    e.stopPropagation();
    if (
      this.state.visible &&
      !this.props.disabled &&
      !this.toggleContainer.current.contains(e.target) &&
      !e.target.classList.contains(`${this.props.prefixCls}-option-disabled`)
    ) {
      this.setState({ visible: false });
      this.props.onPanelVisibleChange(false);
    }
  };
  getWrapperBounding = () => {
    const {
      width,
      height,
      top,
      left
    } = this.triggerWrapper.current.getBoundingClientRect();
    const {
      height: wrapperHeight
    } = this.wrapper.current.getBoundingClientRect();

    const { scrollX, scrollY } = window;

    const positions = {
      top: {
        top: top + scrollY - wrapperHeight - 10,
        left: left + scrollX,
        width
      },
      bottom: {
        top: top + height + scrollY,
        left: left + scrollX,
        width
      }
    };
    return positions[this.props.position];
  };

  setWrapperBounding(cb = () => {}) {
    const { left, top, width } = this.getWrapperBounding();
    this.setState({ left, top, width }, cb);
  }

  onResizeHandler = debounce(() => {
    this.setWrapperBounding();
  }, 500);

  onClear = () => {
    this.setState({ selectedValue: "", visible: false });
    this.props.onChange("");
  };

  render() {
    const { visible, left, top, width } = this.state;
    const {
      prefixCls,
      className,
      disabled,
      placeholder,
      children,
      getPopupContainer,
      size,
      style,
      allowClear,
      notFoundContent,
      loading,
      popupContainerClassName,
      labelInValue, //eslint-disable-line
      onPanelVisibleChange, //eslint-disable-line
      ...attr
    } = this.props;

    return (
      <div
        className={cls(`${prefixCls}`, className)}
        {...attr}
        style={style}
        ref={this.toggleContainer}
      >
        <div
          className={cls(`${prefixCls}-inner`, {
            [`${prefixCls}-active`]: visible
          })}
          ref={this.triggerWrapper}
        >
          <Input
            disabled={disabled}
            readonly
            placeholder={placeholder}
            className={cls(`${prefixCls}-input`)}
            value={this.selectedValue}
            onClick={this.onClickHandler}
            size={size}
            style={{
              width: style && style.width
            }}
            suffix={
              loading ? (
                <LoadingIcon className={`${prefixCls}-loading`} />
              ) : (
                <DownIcon className={`${prefixCls}-arrow`} />
              )
            }
            allowClear={allowClear}
            onClear={this.onClear}
          />
        </div>
        {createPortal(
          <div
            className={cls(`${prefixCls}-content`, popupContainerClassName, {
              [`${prefixCls}-open`]: visible,
              [`${prefixCls}-close`]: !visible,
              ["cuke-ui-no-animate"]: visible === null
            })}
            ref={this.wrapper}
            style={{
              width,
              left,
              top
            }}
          >
            {children
              ? React.Children.map(children, (element, index) => {
                  return React.cloneElement(element, {
                    key: index,
                    selectedValue: this.selectedValue,
                    onChange: this.onChange
                  });
                })
              : notFoundContent}
          </div>,
          getPopupContainer()
        )}
      </div>
    );
  }
  componentWillUnmount() {
    window.removeEventListener("click", this.onClickOutsideHandler, false);
    window.removeEventListener("resize", this.onResizeHandler);
  }
  componentDidMount() {
    window.addEventListener("click", this.onClickOutsideHandler, false);
    window.addEventListener("resize", this.onResizeHandler);
    if (this.props.visible) {
      this.setWrapperBounding();
    }
  }
}
