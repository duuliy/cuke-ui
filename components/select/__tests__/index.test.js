import React from "react";
import assert from "power-assert";
import { render, shallow } from "enzyme";
import toJson from "enzyme-to-json";
import Select from "../index";
import Empty from "../../empty";
import Button from "../../button";

describe("<Select/>", () => {
  it("should render a <Select/> components", () => {
    const wrapper = render(
      <Select>
        <Select.Option value="黄瓜">黄瓜</Select.Option>
        <Select.Option value="茄子">茄子</Select.Option>
        <Select.Option value="番茄">番茄</Select.Option>
      </Select>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it("should render custom size", () => {
    const wrapper = render(
      <div>
        <Select placeholder="请选择" size="small">
          <Select.Option value="黄瓜">黄瓜</Select.Option>
          <Select.Option value="茄子">茄子</Select.Option>
          <Select.Option value="番茄">番茄</Select.Option>
        </Select>
        <Select placeholder="请选择">
          <Select.Option value="黄瓜">黄瓜</Select.Option>
          <Select.Option value="茄子">茄子</Select.Option>
          <Select.Option value="番茄">番茄</Select.Option>
        </Select>
        <Select placeholder="请选择" size="large">
          <Select.Option value="黄瓜">黄瓜</Select.Option>
          <Select.Option value="茄子">茄子</Select.Option>
          <Select.Option value="番茄">番茄</Select.Option>
        </Select>
      </div>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should find cuke-select classnames", () => {
    const wrapper = shallow(
      <Select>
        <Select.Option value="黄瓜">黄瓜</Select.Option>
        <Select.Option value="茄子">茄子</Select.Option>
        <Select.Option value="番茄">番茄</Select.Option>
      </Select>
    );
    assert(wrapper.find(".cuke-select").length >= 1);
  });

  it("should cannot click when disabled", () => {
    const onChange = jest.fn();
    const wrapper = shallow(
      <Select disabled onChange={onChange}>
        <Select.Option value="黄瓜">黄瓜</Select.Option>
        <Select.Option value="茄子">茄子</Select.Option>
        <Select.Option value="番茄">番茄</Select.Option>
      </Select>
    );
    wrapper
      .find(Select.Option)
      .at(0)
      .simulate("click");
    expect(onChange).not.toHaveBeenCalled();
  });

  it("should can trigger click", () => {
    const onChange = jest.fn();
    const wrapper = shallow(
      <Select onChange={onChange}>
        <Select.Option value="黄瓜">黄瓜</Select.Option>
        <Select.Option value="茄子">茄子</Select.Option>
        <Select.Option value="番茄">番茄</Select.Option>
      </Select>
    );

    wrapper
      .find(Select.Option)
      .at(0)
      .simulate("click");
    expect(onChange).not.toHaveBeenCalled();
  });

  it("should render init default value", () => {
    const wrapper = shallow(
      <Select defaultValue="黄瓜">
        <Select.Option value="黄瓜">黄瓜</Select.Option>
        <Select.Option value="茄子">茄子</Select.Option>
        <Select.Option value="番茄">番茄</Select.Option>
      </Select>
    );
    wrapper.find(".cuke-select").simulate("click");
    assert(wrapper.state().selectedValue === "黄瓜");
  });

  it("should render init value", () => {
    const wrapper = shallow(
      <Select value="黄瓜">
        <Select.Option value="黄瓜">黄瓜</Select.Option>
        <Select.Option value="茄子">茄子</Select.Option>
        <Select.Option value="番茄">番茄</Select.Option>
      </Select>
    );
    wrapper.find(".cuke-select").simulate("click");
    assert(wrapper.state().selectedValue === "黄瓜");
  });

  it("should trigger onChange when clicked", () => {
    const onChange = jest.fn();
    const wrapper = shallow(
      <Select.Option value="茄子" onChange={onChange}>
        茄子
      </Select.Option>
    );
    wrapper.find(".cuke-select-option").simulate("click");
    expect(onChange).toHaveBeenCalled();
  });

  it("should window cannot trigger click event when un mount", () => {
    const onClick = jest.fn();
    const wrapper = shallow(
      <Select value="黄瓜">
        <Select.Option value="黄瓜">黄瓜</Select.Option>
        <Select.Option value="茄子">茄子</Select.Option>
        <Select.Option value="番茄">番茄</Select.Option>
      </Select>
    );
    wrapper.unmount();
    window.onclick = () => onClick;
    expect(onClick).not.toHaveBeenCalled();
  });

  it.skip("should panel cannot hidden click event when window click", () => {
    const wrapper = shallow(
      <Select value="黄瓜">
        <Select.Option value="黄瓜">黄瓜</Select.Option>
        <Select.Option value="茄子">茄子</Select.Option>
        <Select.Option value="番茄">番茄</Select.Option>
      </Select>
    );

    window.onclick();
    assert(wrapper.state().visible === null);
    wrapper.find(".cuke-select-input").simulate("click");
    assert(wrapper.state().visible === true);
    window.onclick();
    assert(wrapper.state().visible === true);
  });

  it.skip("should trigger panel visible change", () => {
    const onPanelVisibleChange = jest.fn();
    const wrapper = shallow(
      <Select value="黄瓜">
        <Select.Option value="黄瓜">黄瓜</Select.Option>
        <Select.Option value="茄子">茄子</Select.Option>
        <Select.Option value="番茄">番茄</Select.Option>
      </Select>
    );
    wrapper.find(".cuke-select-input").simulate("click");
    setTimeout(() => {
      expect(onPanelVisibleChange).toHaveBeenCalled();
    });
  });

  it.skip("should cannot trigger panel visible change when disabled groups", () => {
    const onPanelVisibleChange = jest.fn();
    const wrapper = shallow(
      <Select value="黄瓜">
        <Select.Option value="黄瓜">黄瓜</Select.Option>
        <Select.Option value="茄子">茄子</Select.Option>
        <Select.Option value="番茄">番茄</Select.Option>
      </Select>
    );
    wrapper.find(".cuke-select-input").simulate("click");
    expect(onPanelVisibleChange).not.toHaveBeenCalled();
  });

  it("should render not found content when option is Empty", () => {
    const wrapper = shallow(<Select value="黄瓜" />);
    assert(wrapper.find(Empty).length === 1);
    assert(wrapper.find(".cuke-select-option").length === 0);
  });

  it("should render custom not found content", () => {
    const wrapper = shallow(
      <Select value="黄瓜" notFoundContent={<Button>111</Button>} />
    );
    assert(wrapper.find(Button).length === 1);
  });

  it("should find custom popup container class name ", () => {
    const wrapper = shallow(
      <Select value="黄瓜" popupContainerClassName="test" />
    );
    assert(wrapper.find(".test").length === 1);
  });

  it("should find loading icon", () => {
    const wrapper = render(
      <Select value="黄瓜" loading>
        <Select.Option value="黄瓜">123</Select.Option>
        <Select.Option value="茄子">123</Select.Option>
        <Select.Option value="番茄">123</Select.Option>
      </Select>
    );
    expect(wrapper.find(".cuke-select-loading")).toHaveLength(1);
  });

  it("should find loading icon", () => {
    const wrapper = shallow(
      <Select value="黄瓜" loading>
        <Select.Option value="黄瓜">123</Select.Option>
        <Select.Option value="茄子">123</Select.Option>
        <Select.Option value="番茄">123</Select.Option>
      </Select>
    );

    expect(wrapper.state().selectedValue).toEqual("黄瓜");
    wrapper.setProps({ value: "番茄" });
    expect(wrapper.state().selectedValue).toEqual("番茄");
  });
});
