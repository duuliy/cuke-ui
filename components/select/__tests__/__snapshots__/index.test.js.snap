// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select/> should render a <Select/> components 1`] = `
<div
  class="cuke-select"
  position="bottom"
>
  <div
    class="cuke-select-inner"
  >
    <div
      class="cuke-input-wrapper cuke-input-has-suffix"
    >
      <input
        class="cuke-input cuke-select-input"
        placeholder=""
        readonly=""
        type="text"
        value=""
      />
      <span
        class="cuke-input-suffix"
      >
        <svg
          class="cuke-select-arrow"
          fill="none"
          height="1em"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1em"
        >
          <polyline
            points="6 9 12 15 18 9"
          />
        </svg>
      </span>
    </div>
  </div>
  <div
    class="cuke-select-content cuke-select-close cuke-ui-no-animate"
    style="width:0;left:0;top:0"
  >
    <div
      class="cuke-select-option"
    >
      黄瓜
    </div>
    <div
      class="cuke-select-option"
    >
      茄子
    </div>
    <div
      class="cuke-select-option"
    >
      番茄
    </div>
  </div>
</div>
`;

exports[`<Select/> should render custom size 1`] = `
<div>
  <div
    class="cuke-select"
    position="bottom"
  >
    <div
      class="cuke-select-inner"
    >
      <div
        class="cuke-input-wrapper cuke-input-has-suffix"
      >
        <input
          class="cuke-input cuke-select-input cuke-input-small"
          placeholder="请选择"
          readonly=""
          type="text"
          value=""
        />
        <span
          class="cuke-input-suffix"
        >
          <svg
            class="cuke-select-arrow"
            fill="none"
            height="1em"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="1em"
          >
            <polyline
              points="6 9 12 15 18 9"
            />
          </svg>
        </span>
      </div>
    </div>
    <div
      class="cuke-select-content cuke-select-close cuke-ui-no-animate"
      style="width:0;left:0;top:0"
    >
      <div
        class="cuke-select-option"
      >
        黄瓜
      </div>
      <div
        class="cuke-select-option"
      >
        茄子
      </div>
      <div
        class="cuke-select-option"
      >
        番茄
      </div>
    </div>
  </div>
  <div
    class="cuke-select"
    position="bottom"
  >
    <div
      class="cuke-select-inner"
    >
      <div
        class="cuke-input-wrapper cuke-input-has-suffix"
      >
        <input
          class="cuke-input cuke-select-input"
          placeholder="请选择"
          readonly=""
          type="text"
          value=""
        />
        <span
          class="cuke-input-suffix"
        >
          <svg
            class="cuke-select-arrow"
            fill="none"
            height="1em"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="1em"
          >
            <polyline
              points="6 9 12 15 18 9"
            />
          </svg>
        </span>
      </div>
    </div>
    <div
      class="cuke-select-content cuke-select-close cuke-ui-no-animate"
      style="width:0;left:0;top:0"
    >
      <div
        class="cuke-select-option"
      >
        黄瓜
      </div>
      <div
        class="cuke-select-option"
      >
        茄子
      </div>
      <div
        class="cuke-select-option"
      >
        番茄
      </div>
    </div>
  </div>
  <div
    class="cuke-select"
    position="bottom"
  >
    <div
      class="cuke-select-inner"
    >
      <div
        class="cuke-input-wrapper cuke-input-has-suffix"
      >
        <input
          class="cuke-input cuke-select-input cuke-input-large"
          placeholder="请选择"
          readonly=""
          type="text"
          value=""
        />
        <span
          class="cuke-input-suffix"
        >
          <svg
            class="cuke-select-arrow"
            fill="none"
            height="1em"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="1em"
          >
            <polyline
              points="6 9 12 15 18 9"
            />
          </svg>
        </span>
      </div>
    </div>
    <div
      class="cuke-select-content cuke-select-close cuke-ui-no-animate"
      style="width:0;left:0;top:0"
    >
      <div
        class="cuke-select-option"
      >
        黄瓜
      </div>
      <div
        class="cuke-select-option"
      >
        茄子
      </div>
      <div
        class="cuke-select-option"
      >
        番茄
      </div>
    </div>
  </div>
</div>
`;
